### PR TITLE
Added format, language and relation mods fields

### PR DESCRIPTION
**Pull request by kjallen**
This adds the MODS element creation for the following fields:

- `ProjectItem.relation`
- `ProjectItem.format`
- `Language`
- `Description`

The tests could be improved, `Description` is the most complex case where based on the type and/or string value, different attributes or in some cases, such as `abstract`, different elements, are created.

- Review test cases, feedback on improvements or follow up tickets for additional
- Review code for additional exception handling that should be required
- Test with a few objects to confirm MODS is in fact generated using steps below

Pick an item:
http://localhost:8000/item/2526

Open shell:
`$ docker-compose exec django python manage.py shell`

Run:
```
>>> from oh_staff_ui.classes.OralHistoryMods import OralHistoryMods
>>> from oh_staff_ui.models import ProjectItem
>>> pi = ProjectItem.objects.get(id=2526)
>>> ohm = OralHistoryMods(pi)
>>> ohm.populate_fields()
>>> ohm.serializeDocument()
b'<?xml version=\'1.0\' encoding=\'UTF-8\'?>\n<mods:mods xmlns:mods="http://www.loc.gov/mods/v3"><mods:titleInfo><mods:title>Interview of Paul Pickowicz</mods:title></mods:titleInfo><mods:identifier>21198/zz002kppdz</mods:identifier><mods:language><mods:languageTerm>eng</mods:languageTerm></mods:language><mods:physicalDescription><mods:extent>5.25 hrs.</mods:extent></mods:physicalDescription><mods:note displayLabel="Processing of Interview" type="processinterview">The transcript is a verbatim transcription of the recording as transcribed by a machine transcribing program and audit-edited by the interviewer. The interviewee was given an opportunity to review the transcript and made a number of corrections and additions. The transcript may thus differ slightly from the audio recording because of the changes the interviewee made at the time of their review.</mods:note><mods:note displayLabel="Supporting Documents" type="supportingdocuments">Records relating to the interview are located in the office of the UCLA Library\xe2\x80\x99s Center for Oral History Research.</mods:note><mods:note displayLabel="Interviewer Background and Preparation" type="interviewerhistory">The interview was conducted by Xi Chen, MA, MLS, Sally T. WongAvery Librarian of Chinese Studies and East Asia Collection Strategist, University of California, San Diego. The interviewer prepared for the interview by reading Paul G. Pickowicz\xe2\x80\x99s A Sensational Encounter with High Socialist China; Pickowicz\xe2\x80\x99s Chinese Village, Socialist State; Pickowicz\xe2\x80\x99s China on Film: A Century of Exploration, Confrontation, and Controversy; Committee of Concerned Asian Scholars\xe2\x80\x99 China! Inside the People\xe2\x80\x99s Republic; Hanchao Lu\xe2\x80\x99s Versatility, Interdisciplinarity, and Academic Collaboration: Paul Pickowicz\xe2\x80\x99s Insights on Chinese Studies.</mods:note><mods:note displayLabel="Persons Present" type="personpresent">Pickowicz and Chen.</mods:note><mods:note type="biographical" displayLabel="Biographical Information">American historian of modern China and Distinguished Professor of History and Chinese Studies at University of California, San Diego.</mods:note><mods:note displayLabel="Place Conducted" type="place">The interviews were conducted in person at the Geisel Library, University of California, San Diego.</mods:note></mods:mods>'
```
Beautified for reading:

```
<?xml version=\'1.0\' encoding=\'UTF-8\'?>\n
<mods:mods
	xmlns:mods="http://www.loc.gov/mods/v3">
	<mods:titleInfo>
		<mods:title>Interview of Paul Pickowicz</mods:title>
	</mods:titleInfo>
	<mods:identifier>21198/zz002kppdz</mods:identifier>
	<mods:language>
		<mods:languageTerm>eng</mods:languageTerm>
	</mods:language>
	<mods:physicalDescription>
		<mods:extent>5.25 hrs.</mods:extent>
	</mods:physicalDescription>
	<mods:note displayLabel="Processing of Interview" type="processinterview">The transcript is a verbatim transcription of the recording as transcribed by a machine transcribing program and audit-edited by the interviewer. The interviewee was given an opportunity to review the transcript and made a number of corrections and additions. The transcript may thus differ slightly from the audio recording because of the changes the interviewee made at the time of their review.</mods:note>
	<mods:note displayLabel="Supporting Documents" type="supportingdocuments">Records relating to the interview are located in the office of the UCLA Library\xe2\x80\x99s Center for Oral History Research.</mods:note>
	<mods:note displayLabel="Interviewer Background and Preparation" type="interviewerhistory">The interview was conducted by Xi Chen, MA, MLS, Sally T. WongAvery Librarian of Chinese Studies and East Asia Collection Strategist, University of California, San Diego. The interviewer prepared for the interview by reading Paul G. Pickowicz\xe2\x80\x99s A Sensational Encounter with High Socialist China; Pickowicz\xe2\x80\x99s Chinese Village, Socialist State; Pickowicz\xe2\x80\x99s China on Film: A Century of Exploration, Confrontation, and Controversy; Committee of Concerned Asian Scholars\xe2\x80\x99 China! Inside the People\xe2\x80\x99s Republic; Hanchao Lu\xe2\x80\x99s Versatility, Interdisciplinarity, and Academic Collaboration: Paul Pickowicz\xe2\x80\x99s Insights on Chinese Studies.</mods:note>
	<mods:note displayLabel="Persons Present" type="personpresent">Pickowicz and Chen.</mods:note>
	<mods:note type="biographical" displayLabel="Biographical Information">American historian of modern China and Distinguished Professor of History and Chinese Studies at University of California, San Diego.</mods:note>
	<mods:note displayLabel="Place Conducted" type="place">The interviews were conducted in person at the Geisel Library, University of California, San Diego.</mods:note>
</mods:mods>
```

Known issue: While the characters I have seen are validating, I don't think the output is what would be expected on the public site, there are some quote characters (forward-tick? acute accent? this character: `’`). A decision will need to be made on the source data, replacing these and/or swapping for `&apos`

Revisions: 

- Tests : Additional description added with invalid XML characters, should still validate, confirms escaping of characters
- Similar behaving qualifiers grouped together in conditional to improve DRY

### CODE REVIEW

File: oh_staff_ui/tests.py [hunk 1]
```diff
@@ -889,6 +901,27 @@ def setUpTestData(cls):
...
+        # adminnote = 1
+        Description.objects.create(
+            item=cls.interview_item,
+            value="Abstract element",
+            type_id=1,
+        )
+        # adminnote = 4
+        Description.objects.create(
+            item=cls.interview_item,
+            value="Processing of interview:Process interview description as admin note",
```
Comment (akohler) [inline oh_staff_ui/tests.py:915]: If diacritics / special characters are a concern, add a Description with some of them to test the output.

**Review by akohler — CHANGES_REQUESTED** [review]
@kjallen Overall this looks like a good start to me.  The description-handling code should be simplified & improved in a few ways, see individual comments for those.

The tests pass for me.  Per your note about special characters, I tested manually with http://127.0.0.1:8000/item/2492 which has `’` (U+2019 RIGHT SINGLE QUOTATION MARK) in "Library’s".  The serialized XML properly converted that to the correct UTF-8, `\xe2\x80\x99`.  That probably would be a good test case.

If the public site doesn't handle proper UTF-8, the public site needs to be fixed :) - this all seems correct on our end.

Exception handling: so far I don't see any obvious need.  I see that the base `MODS` class provides list-like members for `self.notes` and `self.languages`, so those can safely be appended/extended without worry.  Our database models ensure that `title` and `ark` will always exist.  So I'd focus on differences between our model and the MODS requirements - like the `format` issue you mentioned yesterday - and make sure those are handled in ways that are appropriate.

File: oh_staff_ui/classes/OralHistoryMods.py [hunk 2]
```diff
@@ -12,11 +17,134 @@ def __init__(self, project_item):
...
+        descriptions = Description.objects.filter(item=self._item)
+
+        for desc in descriptions:
+            desc_type = desc.type.type.lower()
+
+            if desc_type == "abstract":
+                self.create_abstract()
+                self.abstract.text = desc.value
+
+            elif desc_type in ["not qualified", "caption", "note"]:
```
[the remark this answers is not in the document]
Comment (kjallen) [reply oh_staff_ui/classes/OralHistoryMods.py:61]: Removed upon review, as there is no functional behavioral difference between these qualifiers and the default else case.

**Review by akohler — APPROVED** [review]

File: oh_staff_ui/classes/OralHistoryMods.py [hunk 3]
```diff
@@ -12,11 +16,100 @@ def __init__(self, project_item):
...
+
+    def _parse_adminnote(self, desc_value: str) -> mods.Note:
+        note_components = desc_value.split(":", 1)
+
+        if len(note_components) <= 1:
+            return mods.Note(text=desc_value)
+
+        else:
+            admin_note_type = note_components[0].lower()
+            admin_note_value = note_components[1]
```
Comment (akohler) [inline oh_staff_ui/classes/OralHistoryMods.py:94]: Does this need a `.strip()`?  Many (most?) notes have a space after the colon, like "Persons Present: Harris Hayes and Collings".  As-is, the `mods.Note.text` will be " Harris Hayes and Collings".

File: oh_staff_ui/classes/OralHistoryMods.py [hunk 4]
```diff
@@ -12,11 +16,100 @@ def __init__(self, project_item):
...
+                    type="interviewerhistory",
+                    text=admin_note_value,
+                )
+
+            elif admin_note_type == "processing of interview":
+                return mods.Note(
+                    label="Processing of Interview",
+                    type="processinterview",
+                    text=admin_note_value,
+                )
```
Comment (akohler) [inline oh_staff_ui/classes/OralHistoryMods.py:115]: There's no `else`, so only admin notes with 1 of these 3 types (or no type) will be added.  I checked our data, and of our 582 description records with type=adminnote:
* 288 have no colon, so would return the full value; that's fine
* 294 have a colon, so would get `admin_note_type` checked
* only 1 of those 294 would return a value, for "supporting documents"

Example: http://127.0.0.1:8000/item/9603 (21198/zz002kc9zw).

See https://gist.github.com/akohler/505cb62e2dc51e0320f3983d85ca9201 for the code and output I get.

File: oh_staff_ui/classes/OralHistoryMods.py [hunk 5]
```diff
@@ -12,11 +16,100 @@ def __init__(self, project_item):
...
+            admin_note_value = note_components[1]
+
+            if admin_note_type == "supporting documents":
+                return mods.Note(
+                    label="Supporting Documents",
+                    type="supportingdocuments",
+                    text=admin_note_value,
+                )
+
+            elif admin_note_type == "interviewer background and preperation":
```
Comment (akohler) [inline oh_staff_ui/classes/OralHistoryMods.py:103]: Typo: `preperation` should be `preparation` here and line 105.

**Review by akohler — CHANGES_REQUESTED** [review]
@kjallen 2 very minor changes: a typo in string literals, and may need to strip whitespace from a value.

Maybe a larger problem, maybe irrelevant: `_parse_admin()` currently will not return values for about half of the `adminnote` descriptions.  That may be appropriate - we may have really bad / non-standard data - or more cases may need to be added.  Since most of these are "Persons Present:" and "Place:", which have their own description types... and many may be on file-level items, which Jane says should only have table of contents descriptions anyhow... I suspect bad data.  But please take a look; if the coding is correct and this is bad data, please add a comment to this method confirming that these are the only valid admin note types.

File: oh_staff_ui/classes/OralHistoryMods.py (hunk 3 above)
Comment (kjallen) [reply oh_staff_ui/classes/OralHistoryMods.py:94]: Yes, the tableOfContents logic is a little unorthodox and seems only output based on the presence of a file, and should be covered when the PR for file elements are submitted. For reference I have been following:

[DLDataProvider](https://github.com/UCLALibrary/DLDataProvider/blob/81a60ab4fc668590d276a190e1068ed1f4f79cc6/Model/src/edu/ucla/library/digital/jaxb/generateModsXML.java#L724)

File: oh_staff_ui/classes/OralHistoryMods.py (hunk 4 above)
Comment (kjallen) [reply oh_staff_ui/classes/OralHistoryMods.py:115]: You are correct, there should be an `else` upon review of the [original logic](https://github.com/UCLALibrary/DLDataProvider/blob/81a60ab4fc668590d276a190e1068ed1f4f79cc6/Model/src/edu/ucla/library/digital/jaxb/generateModsXML.java#LL250C21-L250C21)

Comment (kjallen) [reply oh_staff_ui/classes/OralHistoryMods.py:115]: Per discussion with @emcaulay - `Description`s of `DescriptionType` with value `adminnote` will not be displayed in the exported MODS record

File: oh_staff_ui/classes/OralHistoryMods.py [hunk 6]
```diff
@@ -12,11 +16,66 @@ def __init__(self, project_item):
...
+                type_value = (
+                    "biographical" if desc_type == "biographicalnote" else desc_type
+                )
+                self.notes.append(
+                    mods.Note(
+                        type=type_value, label=type_labels[desc_type], text=desc.value
+                    )
+                )
+
+            else:
```
Comment (akohler) [inline oh_staff_ui/classes/OralHistoryMods.py:80]: @kjallen I thought we agreed to exclude admin notes?  This still allows them through, via the `else`, as generic `mods.Note` elements.  Using http://127.0.0.1:8000/item/9603 , which has 2 admin notes and one ToC:
```
<?xml version="1.0" encoding="UTF-8"?>
<mods:mods xmlns:mods="http://www.loc.gov/mods/v3">
  <mods:titleInfo>
    <mods:title>SESSION 1 (3/13/2017)</mods:title>
  </mods:titleInfo>
  <mods:identifier>21198/zz002kc9zw</mods:identifier>
  <mods:note>Persons Present: Harris Hayes and Collings</mods:note>
  <mods:note>Place: Ann Harris Hayes's home in Los Angeles</mods:note>
  <mods:note>Early life and education--Becomes a mother--Works as an interior designer--Sells her own art work--Has an interest in studying to become an architect, but is unable to pursue this direction--Begins work at Universal Studios--Benefits from pressure on studios to hire more women--The reporting structure and work routine in the are of production design at the time--Begins to freelance when the Universal art department is dissolved--Stephen J. Cannell Productions--The job duties of a set designer--Begins work as an art director.</mods:note>
</mods:mods>
```

File: oh_staff_ui/tests.py [hunk 7]
```diff
@@ -920,6 +963,40 @@ def test_valid_audio_item_mods(self):
...
+        # Assert MODS still valid as read from string
+        self.assertEqual(ohmods_from_string.is_valid(), True)
+
+        # Assert specific description type properly exists
+        mods_xml = ohmods_from_string.serialize(pretty=True)
+        self.assertTrue(
+            b'<mods:note type="processinterview" displayLabel="Processing of Interview">'
+            in mods_xml
+        )
+
```
Comment (akohler) [inline oh_staff_ui/tests.py:999]: If we are excluding admin notes, there should be a test which confirms the description added at line 913 above is *not* included in the output.

**Review by akohler — CHANGES_REQUESTED** [review]
Admin notes are still being included; I think we agreed they should not be.

File: oh_staff_ui/classes/OralHistoryMods.py (hunk 6 above)
Comment (kjallen) [reply oh_staff_ui/classes/OralHistoryMods.py:80]: Yes, confirming admin should not show up. Reviewing.

Comment (kjallen) [reply oh_staff_ui/classes/OralHistoryMods.py:81]: Added `exclude` filter for `adminnote` and `tableOfContents`
`else` catch all uses generic `mods:note` for `note` type and if new types added in future.

File: oh_staff_ui/tests.py (hunk 7 above)
Comment (kjallen) [reply oh_staff_ui/tests.py:1011]: Additional tests added for generic `note` and absence of `adminnote`

**Review by akohler — APPROVED** [review]

**Comment by akohler** [comment]
@kjallen All looks correct now.  Lovely to no longer see all that description note handling code 👍 
Tests all pass.  I tweaked some tests locally to confirm they failed as expected when I varied the data.
I tested manually with http://127.0.0.1:8000/item/2500 (21198/zz002knx4r), which contains a mix of wanted description types and an admin note.  Output is all as expected.

Merging! Thanks --Andy

**Pull request merged**